src/ast/passes/resource_analyser.cpp

```diff
@@ -289,6 +302,53 @@ void ResourceAnalyser::visit(Call &call)
       resources_.str_buffers++;
   }
 
+  if (call.func == "count" || call.func == "sum" || call.func == "max" ||
```

Comment:
@ryantimwilson Out of curiosity, can we get away with removing all this code in the `visit(Call &call)` function if we add checks for the map key in `visit(Map &map)`, which you have and in `visit(AssignMapStatement &assignment)` ?

Reply:
@jordalgo unfortunately no for 2 reasons:
1. All calls except has_key and delete set `call.map`. But `Visitors::visit(Call &call)` does not recurse into `call.map` AST node: https://github.com/bpftrace/bpftrace/blob/941391743d58a418cc075983760e79bc8732ee95/src/ast/visitors.cpp#L31-L36. And changing the recursion would likely increase allocations for other buffers, so I recommend we don't change that.
2. For has_key and delete, arg0 is the map. However, these functions have multiple arguments and whether the buffer size depends on the additional argument that we can't access within `visit(Map &map)` and in `visit(AssignMapStatement &assignment)`. For example, for `delete(@, 2)`, `map.key_expr` will be set to nullptr but we will call `needMapKeyAllocaton(@, 2)`

I can add comments to explain this.

Reply:
Ok, so what am I missing here. I pulled down this PR and I removed this bit:
```
if (call.func == "count" || call.func == "sum" || call.func == "max" ||
      call.func == "min" || call.func == "avg" || call.func == "stats") {
    Map &map = *call.map;
    if (needMapKeyAllocation(map) && map_key_exceeds_stack_limit(map)) {
      resources_.map_key_buffers++;
      resources_.max_map_key_size = std::max(resources_.max_map_key_size,
                                             get_map_key_size(map));
    }
  } 
```
And I still got the key scratch buffer because the key_expr exists in `visit(AssignMapStatement)`.

Reply:
Oh yea good point. Its because for these functions, they are always called like `@ = count()`, so it will always trigger AssignMapStatement AST node where we already add map key buffer.

I'll remove this and add a comment why they're not here.

Note we need to keep has_key/delete for the issue I mentioned above and lhist/hist because the buffer is specially sized to include log2/linear function pointer.